# **Важно!**

Домашнее задание состоит из нескольких задач, которые вам нужно решить.
*   Баллы выставляются по принципу выполнено/невыполнено.
*   За каждую выполненую задачу вы получаете баллы (количество баллов за задание указано в скобках).

**Инструкция выполнения:** Выполните задания в этом же ноутбуке (места под решения **КАЖДОЙ** задачи обозначены как **#НАЧАЛО ВАШЕГО РЕШЕНИЯ** и **#КОНЕЦ ВАШЕГО РЕШЕНИЯ**)

**Как отправить задание на проверку:** Вам необходимо сохранить ваше решение в данном блокноте и отправить итоговый **файл .IPYNB** на учебной платформе в **стандартную форму сдачи домашнего задания.**

**Срок проверки преподавателем:** домашнее задание проверяется **в течение 3 дней после дедлайна сдачи** с предоставлением обратной связи

# **Прежде чем проверять задания:**

1. Перезапустите **ядро (restart the kernel)**: в меню, выбрать **Ядро (Kernel)**
→ **Перезапустить (Restart)**
2. Затем **Выполнить** **все ячейки (run all cells)**: в меню, выбрать **Ячейка (Cell)**
→ **Запустить все (Run All)**.

После ячеек с заданием следуют ячейки с проверкой **с помощью assert.**

Если в коде есть ошибки, assert выведет уведомление об ошибке.

Если в коде нет ошибок, assert отработает без вывода дополнительной информации.

### **Задание**

В этом задании вам предстоит разобраться с энкодерами трансформеров на примере задач классификации и работы с различными моделями семейства BERT и XLNet. Задание состоит из трёх частей, каждая из которых поможет вам глубже понять работу трансформеров, их обучение и использование для решения задач обработки естественного языка (NLP).

#### Часть 1: Классификация с использованием модели BERT и стандартного цикла обучения (без `Trainer`)
1. В этой части вам необходимо решить задачу текстовой классификации (например, классификация отзывов на положительные и отрицательные) с использованием предобученной модели из семейства BERT (например, `bert-base-uncased`).
2. Используйте стандартный цикл обучения, без использования класса `Trainer`. Это значит, что вы будете самостоятельно определять цикл обучения, включая шаги прямого прохода, вычисления функции потерь, обратного прохода и обновления весов модели.
3. Краткое руководство:
   - Загрузите датасет, подходящий для задачи классификации (например, IMDb или другой).
   - Используйте библиотеку Huggingface `transformers` для загрузки модели и токенизатора.
   - Реализуйте процесс токенизации входных данных и подготовки батчей.
   - Напишите цикл обучения с использованием библиотеки `torch`:
     - Прямой проход (forward pass) через модель.
     - Вычисление функции потерь.
     - Обратный проход (backward pass) и обновление весов.
   - Оцените точность (accuracy) модели на тестовой выборке.

#### Часть 2: Классификация с кастомной "головой" на базе BERT и использованием класса `Trainer`
1. В этой части вам необходимо решить задачу классификации с использованием модели BERT, но с добавлением кастомной "головы" (custom head) — нового слоя или нескольких слоев поверх базовой модели BERT.
2. Описание задачи:
   - В качестве основы возьмите ту же модель `bert-base-uncased`, но добавьте свой собственный классификатор (например, дополнительные линейные слои или слои с нелинейностью).
   - Используйте класс `Trainer` из библиотеки `transformers` для упрощения процесса обучения.
   - Настройте гиперпараметры (шаг обучения, количество эпох и т.д.) и запустите обучение модели.
3. Краткое руководство:
   - Определите кастомную голову, которая будет использовать выходы последнего слоя BERT для классификации.
   - Создайте новый класс, наследующий от `torch.nn.Module`, и включите в него кастомные слои.
   - Настройте процесс обучения с помощью класса `Trainer`:
     - Определите функции потерь и метрики.
     - Укажите параметры оптимизатора и планировщика (scheduler).
   - Оцените производительность модели с кастомной "головой" на тестовой выборке и сравните с результатами первой части.

#### Часть 3: Сравнение метрик BERT и XLNet на задаче токен-классификации
1. В этой части вам необходимо сравнить производительность двух моделей — BERT и XLNet — на задаче токен-классификации (например, задача распознавания именованных сущностей — Named Entity Recognition, NER).
2. Задача:
   - Возьмите готовый датасет для токен-классификации (например, CoNLL-2003 NER Dataset).
   - Используйте модели BERT (`bert-base-cased`) и XLNet (`xlnet-base-cased`) для решения этой задачи.
   - Обучите обе модели с использованием класса `Trainer` и сравните их производительность по метрикам точности (accuracy), полноты (recall), F1-меры.
3. Краткое руководство:
   - Загрузите и предобработайте данные для задачи токен-классификации (например, используя библиотеку `datasets`).
   - Для каждой из моделей настройте обучение с использованием класса `Trainer`.
   - Обучите модели и получите метрики на тестовой выборке.
   - Сравните результаты: какие различия наблюдаются между производительностью BERT и XLNet на задаче токен-классификации? Как это можно объяснить с точки зрения архитектуры моделей?

#### Ожидаемый результат:
- Все задачи решаются в одном Jupyter notebook с подробным описанием процесса решения задачи, кодом и анализом результатов.
- В третьей части задания необходимо сделать выводы по сравнению производительности моделей BERT и XLNet.
- Включите графики метрик и таблицы результатов для наглядности.

#### Ресурсы:
- Библиотеки: `transformers`, `torch`, `datasets`, `numpy`, `pandas`.
- Датасеты: IMDb (или любой другой для текстовой классификации), CoNLL-2003 для токен-классификации.


Рекомендуемые ресурсы:
- Документация Hugging Face Transformers https://huggingface.co/docs/transformers/index
- Документация Hugging Face Datasets https://huggingface.co/docs/datasets/index

---

In [1]:
!pip install evaluate datasets==3.6.0 seqeval

Defaulting to user installation because normal site-packages is not writeable


## Часть 1 ##

### Подготовка датасета. ###

In [2]:
# Настроим токенизатор
from transformers import AutoTokenizer

model_name = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(model_name)

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

In [3]:
# Загрузка датасета
from datasets import load_dataset

raw_dataset = load_dataset("stanfordnlp/imdb") # Возьмем датасет stanfordnlp/imdb
print(raw_dataset)
print(raw_dataset["train"][0])

KeyboardInterrupt: 

In [ ]:
# Процесс токенизации
def tokenize_batch(batch):
    return tokenizer(
        batch["text"],
        padding="max_length",
        truncation=True,
        max_length=128,
    )

tokenized_dataset = raw_dataset.map(tokenize_batch, batched=True)
tokenized_dataset = tokenized_dataset.remove_columns(["text"])
tokenized_dataset = tokenized_dataset.rename_column("label", "labels")
tokenized_dataset.set_format("torch")

In [ ]:
from torch.utils.data import DataLoader

train_dataloader = DataLoader(
    tokenized_dataset["train"], shuffle=True, batch_size=128
)
eval_dataloader = DataLoader(
    tokenized_dataset["test"], batch_size=128
)

### Загрузка модели и настройко компонентов обучения 

In [ ]:
import torch

device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
device

In [ ]:
# Настроим классификатор
from transformers import AutoModelForSequenceClassification

model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=2).to(device)

In [ ]:
from torch.optim import AdamW

optimizer = AdamW(model.parameters(), lr=5e-5)

In [ ]:
from transformers import get_scheduler

num_epochs = 3
num_training_steps = num_epochs * len(train_dataloader)
lr_scheduler = get_scheduler(
    "linear",
    optimizer=optimizer,
    num_warmup_steps=0,
    num_training_steps=num_training_steps,
)
print(num_training_steps)

### Обучение

In [ ]:
from tqdm.auto import tqdm

progress_bar = tqdm(range(num_training_steps))

model.train()
for epoch in range(num_epochs):
    for batch in train_dataloader:
        batch = {k: v.to(device) for k, v in batch.items()}
        outputs = model(**batch)
        loss = outputs.loss
        loss.backward()

        optimizer.step()
        lr_scheduler.step()
        optimizer.zero_grad()
        progress_bar.update(1)

In [ ]:
import evaluate
accuracy = evaluate.load("accuracy")
model.eval()
for batch in eval_dataloader:
    batch = {k: v.to(device) for k, v in batch.items()}
    with torch.no_grad():
        outputs = model(**batch)

    logits = outputs.logits
    predictions = torch.argmax(logits, dim=-1)
    accuracy.add_batch(predictions=predictions, references=batch["labels"])

print(accuracy.compute())

Итого, на тестовой выборке показатель accuracy составил 0.88932 для модели bert-base-uncased, обученной на 3-х эпохах

## Часть 2

### Классификатор с кастомной головой

Напишем класс для классификатор с кастомной головой на базе bert-base-uncased, добавив один линейный слой

In [ ]:
import torch
import torch.nn as nn
from transformers import AutoConfig, AutoModel
from torch.nn import CrossEntropyLoss

class TransformerForSequenceClassification(nn.Module):
    def __init__(self, config_name: str, num_labels: int):
        super().__init__()
        config = AutoConfig.from_pretrained(config_name, num_labels=num_labels)
        self.encoder = AutoModel.from_pretrained(config_name, config=config)
        self.dropout = nn.Dropout(config.hidden_dropout_prob)
        self.classifier = nn.Linear(config.hidden_size, num_labels)
        self.loss_fct = CrossEntropyLoss()

    def forward(self, input_ids, attention_mask=None, token_type_ids=None, labels=None):
        outputs = self.encoder(
            input_ids=input_ids,
            attention_mask=attention_mask,
            token_type_ids=token_type_ids,
            return_dict=True,
        )
        cls_state = outputs.last_hidden_state[:, 0, :]
        cls_state = self.dropout(cls_state)
        logits = self.classifier(cls_state)

        loss = None
        if labels is not None:
            loss = self.loss_fct(logits.view(-1, self.classifier.out_features), labels.view(-1))

        return (loss, logits) if loss is not None else logits

In [ ]:
model = TransformerForSequenceClassification(model_name, 2).to(device)

### Обучение

Запустим обучение с помощью класса Trainer, позволяющего упростить многие процессы

In [ ]:
from transformers import TrainingArguments

training_args = TrainingArguments(
    output_dir="./results",
    eval_strategy="epoch",
    save_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=128,
    per_device_eval_batch_size=128,
    num_train_epochs=3,
    weight_decay=0.01,
    load_best_model_at_end=True,
    report_to="none",
)

In [ ]:
from transformers import Trainer

trainer = Trainer(
    model,
    training_args,
    train_dataset=tokenized_dataset["train"],
    eval_dataset=tokenized_dataset["test"],
    processing_class=tokenizer,
)

In [ ]:
trainer.train()

TrainOutput(global_step=588, training_loss=0.2591423258489492, metrics={'train_runtime': 1983.9905, 'train_samples_per_second': 37.803, 'train_steps_per_second': 0.296, 'total_flos': 0.0, 'train_loss': 0.2591423258489492, 'epoch': 3.0})

In [ ]:
import evaluate
accuracy = evaluate.load("accuracy")
model.eval()
for batch in eval_dataloader:
    batch = {k: v.to(device) for k, v in batch.items()}
    with torch.no_grad():
        outputs = model(**batch)

    logits = outputs[1]
    predictions = torch.argmax(logits, dim=-1)
    accuracy.add_batch(predictions=predictions, references=batch["labels"])

print(accuracy.compute())

{'accuracy': 0.88696}


Итого, на тестовой выборке показатель accuracy составил 0.88696 для модели кастомной головой, обученной на 3-х эпохах. Показатель стал немного ниже в сравнении с моделью из AutoModelForSequenceClassification (с показателем 0.88932), но не значительно. Скорее всего это связано с тем, что в данном классе используется более сложная архитектура "головы", возможно с несколькими линейными слоями или нелинейностями.

## Задание 3 (XLNET)



### Подготовка датасета
Для работы с датасетом, надо его предобработать, токензировав лейблы

In [ ]:
# Настроим токенизатор
from transformers import AutoTokenizer

# model_name = "bert-base-cased"
model_name = "xlnet-base-cased"
tokenizer = AutoTokenizer.from_pretrained(model_name)

In [ ]:
# Загрузка датасета
from datasets import load_dataset

raw_dataset = load_dataset("eriktks/conll2003")
print(raw_dataset)
print(raw_dataset["train"][0])

DatasetDict({
    train: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 14041
    })
    validation: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 3250
    })
    test: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 3453
    })
})
{'id': '0', 'tokens': ['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.'], 'pos_tags': [22, 42, 16, 21, 35, 37, 16, 21, 7], 'chunk_tags': [11, 21, 11, 12, 21, 22, 11, 12, 0], 'ner_tags': [3, 0, 7, 0, 0, 0, 7, 0, 0]}


In [ ]:
label_all_tokens = True
task = 'ner'
label_list = raw_dataset["train"].features[f"{task}_tags"].feature.names
label_list

['O', 'B-PER', 'I-PER', 'B-ORG', 'I-ORG', 'B-LOC', 'I-LOC', 'B-MISC', 'I-MISC']

In [ ]:
def tokenize_and_align_labels(examples):
    tokenized_inputs = tokenizer(examples["tokens"], truncation=True, is_split_into_words=True)

    labels = []
    for i, label in enumerate(examples[f"{task}_tags"]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids = []
        for word_idx in word_ids:
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                label_ids.append(label[word_idx])
            else:
                label_ids.append(label[word_idx] if label_all_tokens else -100)
            previous_word_idx = word_idx

        labels.append(label_ids)

    tokenized_inputs["labels"] = labels
    return tokenized_inputs

tokenized_dataset = raw_dataset.map(
    tokenize_and_align_labels,
    batched=True,
)

Map:   0%|          | 0/3250 [00:00<?, ? examples/s]

Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


### Настройка модели и параметров обучения

In [ ]:
from transformers import AutoModelForTokenClassification

model = AutoModelForTokenClassification.from_pretrained(model_name, num_labels=len(label_list))

2025-10-07 20:39:01.170717: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-10-07 20:39:01.193244: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2025-10-07 20:39:01.193259: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2025-10-07 20:39:01.194211: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2025-10-07 20:39:01.198202: I tensorflow/core/platform/cpu_feature_guar

In [ ]:
from transformers import TrainingArguments

training_args = TrainingArguments(
    output_dir="./results_3",
    eval_strategy="epoch",
    save_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=32,
    per_device_eval_batch_size=32,
    num_train_epochs=3,
    weight_decay=0.01,
    load_best_model_at_end=True,
    report_to=["tensorboard"],
)

In [ ]:
from transformers import DataCollatorForTokenClassification

data_collator = DataCollatorForTokenClassification(tokenizer)

In [ ]:
import evaluate

metric = evaluate.load("seqeval")

In [ ]:
import numpy as np

def compute_metrics(p):
    predictions, labels = p
    predictions = np.argmax(predictions, axis=2)

    # Remove ignored index (special tokens)
    true_predictions = [
        [label_list[p] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]
    true_labels = [
        [label_list[l] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]

    results = metric.compute(predictions=true_predictions, references=true_labels)
    return {
        "precision": results["overall_precision"],
        "recall": results["overall_recall"],
        "f1": results["overall_f1"],
        "accuracy": results["overall_accuracy"],
    }

### Обучение

In [ ]:
from transformers import Trainer

trainer = Trainer(
    model,
    training_args,
    train_dataset=tokenized_dataset["train"],
    eval_dataset=tokenized_dataset["validation"],
    data_collator=data_collator,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

/tmp/ipykernel_10132/457321303.py:3: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


In [ ]:
trainer.train()

TrainOutput(global_step=1317, training_loss=0.09945637652977181, metrics={'train_runtime': 183.5711, 'train_samples_per_second': 229.464, 'train_steps_per_second': 7.174, 'total_flos': 1346486970567654.0, 'train_loss': 0.09945637652977181, 'epoch': 3.0})

## Задание 3. (BERT)

### Подготовка датасета

In [1]:
# Настроим токенизатор
from transformers import AutoTokenizer

model_name = "bert-base-cased"
tokenizer = AutoTokenizer.from_pretrained(model_name)

In [2]:
# Загрузка датасета
from datasets import load_dataset

raw_dataset = load_dataset("eriktks/conll2003")
print(raw_dataset)
print(raw_dataset["train"][0])

DatasetDict({
    train: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 14041
    })
    validation: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 3250
    })
    test: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 3453
    })
})
{'id': '0', 'tokens': ['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.'], 'pos_tags': [22, 42, 16, 21, 35, 37, 16, 21, 7], 'chunk_tags': [11, 21, 11, 12, 21, 22, 11, 12, 0], 'ner_tags': [3, 0, 7, 0, 0, 0, 7, 0, 0]}


In [3]:
label_all_tokens = True
task = 'ner'
label_list = raw_dataset["train"].features[f"{task}_tags"].feature.names
label_list

['O', 'B-PER', 'I-PER', 'B-ORG', 'I-ORG', 'B-LOC', 'I-LOC', 'B-MISC', 'I-MISC']

In [4]:
def tokenize_and_align_labels(examples):
    tokenized_inputs = tokenizer(examples["tokens"], truncation=True, is_split_into_words=True)

    labels = []
    for i, label in enumerate(examples[f"{task}_tags"]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids = []
        for word_idx in word_ids:
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                label_ids.append(label[word_idx])
            else:
                label_ids.append(label[word_idx] if label_all_tokens else -100)
            previous_word_idx = word_idx

        labels.append(label_ids)

    tokenized_inputs["labels"] = labels
    return tokenized_inputs

tokenized_dataset = raw_dataset.map(
    tokenize_and_align_labels,
    batched=True,
)

Map:   0%|          | 0/14041 [00:00<?, ? examples/s]

Map:   0%|          | 0/3250 [00:00<?, ? examples/s]

Map:   0%|          | 0/3453 [00:00<?, ? examples/s]

### Настройка модели и параметров обучения

In [5]:
from transformers import AutoModelForTokenClassification

model = AutoModelForTokenClassification.from_pretrained(model_name, num_labels=len(label_list))

2025-10-07 23:21:53.695596: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-10-07 23:21:54.143689: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2025-10-07 23:21:54.143735: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2025-10-07 23:21:54.218713: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2025-10-07 23:21:54.364761: I tensorflow/core/platform/cpu_feature_guar

model.safetensors:   0%|          | 0.00/436M [00:00<?, ?B/s]

Some weights of BertForTokenClassification were not initialized from the model checkpoint at bert-base-cased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [6]:
from transformers import TrainingArguments

training_args = TrainingArguments(
    output_dir="./results_3",
    eval_strategy="epoch",
    save_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=32,
    per_device_eval_batch_size=32,
    num_train_epochs=3,
    weight_decay=0.01,
    load_best_model_at_end=True,
    report_to=["tensorboard"],
)

In [7]:
from transformers import DataCollatorForTokenClassification

data_collator = DataCollatorForTokenClassification(tokenizer)

In [8]:
import evaluate

metric = evaluate.load("seqeval")

In [9]:
import numpy as np

def compute_metrics(p):
    predictions, labels = p
    predictions = np.argmax(predictions, axis=2)

    # Remove ignored index (special tokens)
    true_predictions = [
        [label_list[p] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]
    true_labels = [
        [label_list[l] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]

    results = metric.compute(predictions=true_predictions, references=true_labels)
    return {
        "precision": results["overall_precision"],
        "recall": results["overall_recall"],
        "f1": results["overall_f1"],
        "accuracy": results["overall_accuracy"],
    }

### Обучение

In [10]:
from transformers import Trainer

trainer = Trainer(
    model,
    training_args,
    train_dataset=tokenized_dataset["train"],
    eval_dataset=tokenized_dataset["validation"],
    data_collator=data_collator,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

/tmp/ipykernel_23943/457321303.py:3: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


In [11]:
trainer.train()

TrainOutput(global_step=1317, training_loss=0.10604328700060543, metrics={'train_runtime': 140.0699, 'train_samples_per_second': 300.728, 'train_steps_per_second': 9.402, 'total_flos': 1164123671696082.0, 'train_loss': 0.10604328700060543, 'epoch': 3.0})